### PR TITLE
*: check last sent snapshot for prepare merge

### LINKED ISSUE #12663: tikv panic: 12475 first index 56 > min_index 55, skip pre merge
### What version of TiKV are you using?
```
[2022/05/25 15:38:59.799 +00:00] [INFO] [lib.rs:79] ["Welcome to TiKV"]
[2022/05/25 15:38:59.799 +00:00] [INFO] [lib.rs:84] ["Release Version:   6.1.0-alpha"]
[2022/05/25 15:38:59.799 +00:00] [INFO] [lib.rs:84] ["Edition:           Community"]
[2022/05/25 15:38:59.799 +00:00] [INFO] [lib.rs:84] ["Git Commit Hash:   1fb8980ccab9ff40c1adc206df52952dab8e8ad8"]
[2022/05/25 15:38:59.799 +00:00] [INFO] [lib.rs:84] ["Git Commit Branch: heads/refs/tags/v6.1.0"]
```

### What operating system and CPU are you using?
`do not care`

### Steps to reproduce
```
run jepsen jepsen-monotonic test with the following options:
{:auto-retry-limit :default,
 :ssh-private-key "jepsen.pem",
 :isolation :repeatable-read,
 :concurrency 15,
 :predicate-read false,
 :tarball-url
 "http://minio.pingcap.net:9000/tp-team/tests/jepsen/tidb-v6.1.0-pre-amd64.tar.gz",
 :txn-mode "mixed",
 :force-reinstall true,
 :update-in-place false,
 :ssh
 {:username "root",
  :password "root",
  :strict-host-key-checking false,
  :private-key-path "jepsen.pem"},
 :nemesis-long-recovery false,
 :follower-read false,
 :nemesis
 {:interval 10,
  :schedules true,
  :partition-pd-leader true,
  :partition-half true,
  :partition-ring true},
 :nodes
 ["node-0.node-peer.jepsen-tps-840631-1-732"
  "node-1.node-peer.jepsen-tps-840631-1-732"
  "node-2.node-peer.jepsen-tps-840631-1-732"
  "node-3.node-peer.jepsen-tps-840631-1-732"
  "node-4.node-peer.jepsen-tps-840631-1-732"],
 :test-count 1,
 :read-lock "FOR UPDATE",
 :use-index true,
 :os #object[tidb.core.Image 0x19fec3d6 "tidb.core.Image@19fec3d6"],
 :time-limit 300,
 :recovery-time 10,
 :version "master",
 :auto-retry :default,
 :workload :monotonic,
 :init-txn-sql
 ("set @@tidb_enable_async_commit = 0, @@tidb_enable_1pc = 0"
  "set @@tidb_enable_async_commit = 1, @@tidb_enable_1pc = 0"
  "set @@tidb_enable_async_commit = 1, @@tidb_enable_1pc = 1"),
 :init-sql
 ("set @@tidb_enable_mutation_checker=1, @@tidb_txn_assertion_level=strict")}
```

### What did you expect?

`case run passed.`

### What did happened?
```
case run failed due to tikv panic:
[2022/05/25 15:44:59.596 +00:00] [INFO] [raft.rs:1332] ["[logterm: 11, index: 12, vote: 1621] ignored vote from 1622 [logterm: 11, index: 12]: lease is not expired"] ["msg type"=MsgRequestPreVote] ["remaining ticks"=10] [term=11] [msg_index=12] [msg_term=11] [from=1622] [vote=1621] [log_index=12] [log_term=11] [raft_id=1621] [region_id=1619]
[2022/05/25 15:44:59.599 +00:00] [FATAL] [lib.rs:491] ["[region 1081] 12475 first index 56 > min_index 55, skip pre merge"] [backtrace="   0: tikv_util::set_panic_hook::{{closure}}\n             at /home/jenkins/agent/workspace/build-common/go/src/github.com/pingcap/tikv/components/tikv_util/src/lib.rs:490:18\n   1: std::panicking::rust_panic_with_hook\n             at /rust/toolchains/nightly-2022-02-14-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/std/src/panicking.rs:702:17\n   2: std::panicking::begin_panic_handler::{{closure}}\n             at /rust/toolchains/nightly-2022-02-14-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/std/src/panicking.rs:588:13\n   3: std::sys_common::backtrace::__rust_end_short_backtrace\n             at /rust/toolchains/nightly-2022-02-14-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/std/src/sys_common/backtrace.rs:138:18\n   4: rust_begin_unwind\n             at /rust/toolchains/nightly-2022-02-14-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/std/src/panicking.rs:584:5\n   5: core::panicking::panic_fmt\n             at /rust/toolchains/nightly-2022-02-14-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/core/src/panicking.rs:143:14\n   6: raftstore::store::fsm::apply::ApplyDelegate<EK>::exec_admin_cmd\n   7: raftstore::store::fsm::apply::ApplyDelegate<EK>::exec_raft_cmd\n             at /home/jenkins/agent/workspace/build-common/go/src/github.com/pingcap/tikv/components/raftstore/src/store/fsm/apply.rs:1381:13\n      raftstore::store::fsm::apply::ApplyDelegate<EK>::apply_raft_cmd\n             at /home/jenkins/agent/workspace/build-common/go/src/github.com/pingcap/tikv/components/raftstore/src/store/fsm/apply.rs:1241:41\n   8: raftstore::store::fsm::apply::ApplyDelegate<EK>::process_raft_cmd\n             at /home/jenkins/agent/workspace/build-common/go/src/github.com/pingcap/tikv/components/raftstore/src/store/fsm/apply.rs:1194:39\n   9: raftstore::store::fsm::apply::ApplyDelegate<EK>::handle_raft_entry_normal\n             at /home/jenkins/agent/workspace/build-common/go/src/github.com/pingcap/tikv/components/raftstore/src/store/fsm/apply.rs:1077:20\n      raftstore::store::fsm::apply::ApplyDelegate<EK>::handle_raft_committed_entries\n             at /home/jenkins/agent/workspace/build-common/go/src/github.com/pingcap/tikv/components/raftstore/src/store/fsm/apply.rs:980:43\n  10: raftstore::store::fsm::apply::ApplyFsm<EK>::handle_apply\n             at /home/jenkins/agent/workspace/build-common/go/src/github.com/pingcap/tikv/components/raftstore/src/store/fsm/apply.rs:3362:9\n  11: raftstore::store::fsm::apply::ApplyFsm<EK>::handle_tasks\n             at /home/jenkins/agent/workspace/build-common/go/src/github.com/pingcap/tikv/components/raftstore/src/store/fsm/apply.rs:3643:25\n  12: <raftstore::store::fsm::apply::ApplyPoller<EK> as batch_system::batch::PollHandler<raftstore::store::fsm::apply::ApplyFsm<EK>,raftstore::store::fsm::apply::ControlFsm>>::handle_normal\n             at /home/jenkins/agent/workspace/build-common/go/src/github.com/pingcap/tikv/components/raftstore/src/store/fsm/apply.rs:3892:9\n  13: batch_system::batch::Poller<N,C,Handler>::poll\n             at /home/jenkins/agent/workspace/build-common/go/src/github.com/pingcap/tikv/components/batch-system/src/batch.rs:421:27\n  14: raftstore::store::worker::refresh_config::PoolController<N,C,H>::increase_by::{{closure}}\n             at /home/jenkins/agent/workspace/build-common/go/src/github.com/pingcap/tikv/components/raftstore/src/store/worker/refresh_config.rs:76:21\n      std::sys_common::backtrace::__rust_begin_short_backtrace\n             at /rust/toolchains/nightly-2022-02-14-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/std/src/sys_common/backtrace.rs:122:18\n  15: std::thread::Builder::spawn_unchecked_::{{closure}}::{{closure}}\n             at /rust/toolchains/nightly-2022-02-14-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/std/src/thread/mod.rs:498:17\n      <core::panic::unwind_safe::AssertUnwindSafe<F> as core::ops::function::FnOnce<()>>::call_once\n             at /rust/toolchains/nightly-2022-02-14-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/core/src/panic/unwind_safe.rs:271:9\n      std::panicking::try::do_call\n             at /rust/toolchains/nightly-2022-02-14-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/std/src/panicking.rs:492:40\n      std::panicking::try\n             at /rust/toolchains/nightly-2022-02-14-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/std/src/panicking.rs:456:19\n      std::panic::catch_unwind\n             at /rust/toolchains/nightly-2022-02-14-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/std/src/panic.rs:137:14\n      std::thread::Builder::spawn_unchecked_::{{closure}}\n             at /rust/toolchains/nightly-2022-02-14-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/std/src/thread/mod.rs:497:30\n      core::ops::function::FnOnce::call_once{{vtable.shim}}\n             at /rust/toolchains/nightly-2022-02-14-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/core/src/ops/function.rs:227:5\n  16: <alloc::boxed::Box<F,A> as core::ops::function::FnOnce<Args>>::call_once\n             at /rust/toolchains/nightly-2022-02-14-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/alloc/src/boxed.rs:1854:9\n      <alloc::boxed::Box<F,A> as core::ops::function::FnOnce<Args>>::call_once\n             at /rust/toolchains/nightly-2022-02-14-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/alloc/src/boxed.rs:1854:9\n      std::sys::unix::thread::Thread::new::thread_start\n             at /rust/toolchains/nightly-2022-02-14-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/std/src/sys/unix/thread.rs:108:17\n  17: start_thread\n  18: clone\n"] [location=/home/jenkins/agent/workspace/build-common/go/src/github.com/pingcap/tikv/components/raftstore/src/store/fsm/apply.rs:2422] [thread_name=apply-1]

```

### PR DESCRIPTION
**Pull request by BusyJay**
### What is changed and how it works?
Issue Number: Close #12663.

What's Changed:

```commit-message
Guarantee min index of prepare merge larger than the index of last sent snapshot
by recording an approximate last sent snapshot index.
```

### Check List 

Tests 

- Unit test
- Integration test

### Release note 

```release-note
Fix possible panic when source peer catch up logs by snapshot in merge
```

### CODE REVIEW

**Review by Connor1996 — APPROVED** [review]
LGTM

File: components/raftstore/src/store/peer.rs [hunk 1]
```diff
@@ -2020,6 +2029,7 @@ where
                     // prewrites or commits will be just a waste.
                     self.last_urgent_proposal_idx = self.raft_group.raft.raft_log.last_index();
                     self.raft_group.skip_bcast_commit(false);
+                    self.last_sent_snapshot_idx = self.raft_group.raft.raft_log.last_index();
```
Comment (zhangjinpeng1987) [inline components/raftstore/src/store/peer.rs:2032]: Why reset `snapshot_idx` when the role changed to leader?

Comment (BusyJay) [reply components/raftstore/src/store/peer.rs:2032]: Because new leader has no idea what's the snapshot index sent by last leader.

**Review by 5kbpers — APPROVED** [review]

**Pull request merged**